nixos/modules/services/networking/stubby.nix

```diff
@@ -0,0 +1,195 @@
+{ config, lib, pkgs, ...}:
+
+with lib;
+
+let
+  cfg = config.services.stubby;
+
+  stateDir = "/var/lib/stubby";
+  fallbacks = concatMapStringsSep "\n  " (x: "- ${x}") cfg.fallbackProtocols;
+  listeners = concatMapStringsSep "\n  " (x: "- ${x}") cfg.listenAddresses;
+  extraConfig = optionalString (cfg.extraConfig != "") ''
+    ${cfg.extraConfig}
+  '';
+  defaultUpstream = ''
+    - address_data: 145.100.185.15
+      tls_auth_name: "dnsovertls.sinodun.com"
+      tls_pubkey_pinset:
+        - digest: "sha256"
+          value: 62lKu9HsDVbyiPenApnc4sfmSYTHOVfFgL3pyB+cBL4=
+    - address_data: 145.100.185.16
+      tls_auth_name: "dnsovertls1.sinodun.com"
+      tls_pubkey_pinset:
+        - digest: "sha256"
+          value: cE2ecALeE5B+urJhDrJlVFmf38cJLAvqekONvjvpqUA=
+    - address_data: 185.49.141.37
+      tls_auth_name: "getdnsapi.net"
+      tls_pubkey_pinset:
+        - digest: "sha256"
+          value: foxZRnIh9gZpWnl+zEiKa0EJ2rdCGroMWm02gaxSc9Q=
+    - address_data: 2001:610:1:40ba:145:100:185:15
+      tls_auth_name: "dnsovertls.sinodun.com"
+      tls_pubkey_pinset:
+        - digest: "sha256"
+          value: 62lKu9HsDVbyiPenApnc4sfmSYTHOVfFgL3pyB+cBL4=
+    - address_data: 2001:610:1:40ba:145:100:185:16
+      tls_auth_name: "dnsovertls1.sinodun.com"
+      tls_pubkey_pinset:
+        - digest: "sha256"
+          value: cE2ecALeE5B+urJhDrJlVFmf38cJLAvqekONvjvpqUA=
+    - address_data: 2a04:b900:0:100::38
+      tls_auth_name: "getdnsapi.net"
+      tls_pubkey_pinset:
+        - digest: "sha256"
+          value: foxZRnIh9gZpWnl+zEiKa0EJ2rdCGroMWm02gaxSc9Q=
+  '';
+
+  confFile = pkgs.writeText "stubby.yml" ''
+    resolution_type: GETDNS_RESOLUTION_STUB
+    dns_transport_list:
+      ${fallbacks}
+    tls_authentication: ${cfg.authenticationMode}
+    tls_query_padding_blocksize: ${cfg.queryPaddingBlocksize}
+    edns_client_subnet_private: 1
+    idle_timeout: ${cfg.idleTimeout}
+    listen_addresses:
+      ${listeners}
+    round_robin_upstreams: ${cfg.roundRobinUpstreams}
+    ${cfg.extraConfig}
+    upstream_recursive_servers:
+    ${cfg.upstreamServers}
+  '';
```

Comment:
This might be a good case to use the [toYAML](https://github.com/NixOS/nixpkgs/blob/master/lib/generators.nix#L129-L134) function in lib. You could have a single option `services.stubby.config` of type `attrsOf unspecified`, set a default value of `{ resolution_type = "GETDNS_RESOLUTION_STUB"; ... }` and then users can override this config with their own values without removing all others (not 100% sure about that though). Then you can use `toYAML` to generate the config. This has the benefit of not messing up string formatting in any, being easier to enter in Nix, and also being able to override all default settings (the `edns_client_subnet_private: 1` can't be overridden here, and every overridable option would need its own option). `extraConfig` also becomes irrelevant.

Reply:
To still enable the user to pass their own config file without having to convert it to Nix, you could have another option `configFile` which would be set to the generated yaml file from `config` by default.

Reply:
Oh and you could even still keep all the options with the descriptions if you want, using such a `config` option backed by `toYAML` can do that.

Reply:
Hmm, I seem to be having some trouble using `toYAML` for this. Right now, `toYAML` just calls `builtin.toJSON`, which produces a JSON configuration file. The trouble is, `stubby` has an altogether different (and more complex) JSON format that it also accepts, and so they don't line up quite right. It also seems to expect a complete configuration file in the JSON case, rather than merging explicitly set values with the defaults the way it does with YAML.

I will keep experimenting. It's possible I'm still doing something wrong. Are there plans to produce proper YAML with `toYAML` in the future?

Reply:
Well that's not good what they're doing.. Because YAML is a superset of JSON, every JSON is also a valid YAML.

From their source I found that they use the YAML processing when the file ends with `.yml` or `.yaml` though (see https://github.com/getdnsapi/stubby/blob/4b9b0362f97d8db61cd0294c57f583437a5f47b4/src/stubby.c#L280-L281), so I think it should work if you do that.

Reply:
Hmmm, so it still seems something about their parser isn't happy with JSON. I took a confirmed-valid YAML configuration (the default), `yaml.load`ed and `json.dump`ed to JSON, and started the daemon by pointing it at that config file, and I still got the following:

` Apr 08 18:30:28 nixos stubby[14459]: Error parsing config file "/var/lib/stubby/.stubby.yml": Did not update the context`

To be fair, their parser also occasionally has trouble with even valid YAML (IPv6 addresses with leading and trailing colons):

https://github.com/getdnsapi/stubby/issues/18

Reply:
Oh god, they're rolling their own YAML parser and it doesn't even work correctly :/ (yes YAML parsers are hard but JSON compatibility is the least they could do)

Can you file an issue at their repo for this? If they don't fix this soon this module will just have to roll with the inflexible configuration method. I don't think adjusting our `toYAML` function for this is reasonable, as it does indeed work correctly (and since it outputs JSON, Nix can easily parse it again with `builtins.fromJSON`).

Reply:
totally agree regarding the functionality of `toYAML` vs them rolling their own parser. I will open that issue. In the meantime, I add in your suggestions to the previous method.

In any case, thank you for making me aware of  `toYAML` , as it's bound to come in handy sooner than later.